firmware/tracing_firmware/main.rs

```diff
@@ -0,0 +1,178 @@
+#![no_std]
+#![no_main]
+
+use core::arch::{asm, global_asm};
+
+use miralis_abi::{failure, log, setup_binary, success};
+
+setup_binary!(main);
+
+fn enable_mcycle_in_smode() {
+    unsafe {
+        let mcounteren: u32;
+        asm!("csrr {}, mcounteren", out(reg) mcounteren);
+        asm!("csrw mcounteren, {}", in(reg) mcounteren | 1);
+    }
+}
+
+fn main() -> ! {
+    unsafe {
+        asm!("la sp, 0x80700000");
+    }
+
+    enable_mcycle_in_smode();
+
+    let os: usize = operating_system as usize;
+    let trap: usize = _empty_handler as usize;
+    let mpp = 0b1 << 11; // MPP = S-mode
+
+    unsafe {
+        asm!(
+        "li t4, 0xfffffffff",
+        "csrw pmpcfg0, 0xf",   // XRW TOR
+        "csrw pmpaddr0, t4",   // All memory
+        "auipc t4, 0",
+        "addi t4, t4, 24",
+        "csrw mtvec, {mtvec}", // Write mtvec with trap handler
+        "csrw mstatus, {mpp}", // Write MPP of mstatus to S-mode
+        "csrw mepc, {os}",     // Write MEPC
+        "mret",                // Jump to OS
+
+
+        os = in(reg) os,
+        mtvec = in(reg) trap,
+        mpp = in(reg) mpp,
+        out("t4") _,
+        );
+    }
+
+    success();
+}
+
+// —————————————————————————————— Trap Handler —————————————————————————————— //
+
+global_asm!(
+    r#"
+.text
+.align 4
+.global _empty_handler
+_empty_handler:
+    // Skip illegal instruction (pc += 4)
+    csrrw x5, mepc, x5
+    addi x5, x5, 4
+    csrrw x5, mepc, x5
+    // Return back to miralis
+    mret
+"#,
+);
+
+extern "C" {
+    fn _empty_handler();
+}
+
+// —————————————————————————————— Benchmark operating system —————————————————————————————— //
+
+const NB_REPEATS: usize = 1000;
+
+pub fn bubble_sort(arr: &mut [usize; NB_REPEATS]) {
+    let len = arr.len();
+    let mut swapped;
+
+    for i in 0..len {
+        swapped = false;
+
+        for j in 0..len - 1 - i {
+            if arr[j] > arr[j + 1] {
+                arr.swap(j, j + 1);
+                swapped = true;
+            }
+        }
+
+        if !swapped {
+            break;
+        }
+    }
+
+    for i in 1..len {
+        if arr[i - 1] > arr[i] {
+            log::error!("Error in sorting, results aren't reliable");
+            failure();
+        }
+    }
+}
+
+fn operating_system() -> ! {
+    log::info!("Start benchmarking");
+
+    let mut values: [usize; NB_REPEATS] = [0; NB_REPEATS];
+
+    for i in 0..NB_REPEATS {
+        values[i] = trigger_ctx_switch_to_firmware()
+    }
+
+    print_statistics(get_statistics(values));
+
+    success();
+}
+
+fn trigger_ctx_switch_to_firmware() -> usize {
+    let begin: u64;
+    let end: u64;
+
+    unsafe {
+        // Read the `mcycle` register (assuming 64-bit RISC-V)
+        asm!("csrr {}, cycle", out(reg) begin);
```

Comment:
The only thing we need to be sure here is that reading the cycles doesn't trap :)

You were also talking about taking one big measurement of `NB_REPEATS` traps in a row to double check, do you want to do it here too or do we keep that for future work?